RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc

```diff
@@ -0,0 +1,428 @@
+  iEvent.getByLabel(beamSpotLabel_, theBeamSpot);
+
+  // Conversions in miniAOD and AOD have different names,
+  // but the same type, so we use the same handle with different tokens.
+  iEvent.getByLabel(conversionsLabelAOD_, conversions);
+  if( !conversions.isValid() )
+    iEvent.getByLabel(conversionsLabelMiniAOD_, conversions);
+
+  // Make sure everything is retrieved successfully
+  if(! (theBeamSpot.isValid()
+    && conversions.isValid() )
+     )
+    throw cms::Exception("MVA failure: ")
+      << "Failed to retrieve event content needed for this MVA"
+      << std::endl
+      << "Check python MVA configuration file."
+      << std::endl;
+
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
+  if( eleRecoPtr.get() == nullptr )
+    throw cms::Exception("MVA failure: ")
+      << " given particle is expected to be reco::GsfElectron or pat::Electron," << std::endl
+      << " but appears to be neither" << std::endl;
+
+  return fillMVAVariables(eleRecoPtr.get(), conversions, theBeamSpot.product(), theRho);
+}
+
+// A function that should work on both pat and reco objects
+std::vector<float> ElectronMVAEstimatorRun2Fall17::
+fillMVAVariables(const reco::GsfElectron* eleRecoPtr, const edm::Handle<reco::ConversionCollection> conversions, const reco::BeamSpot *theBeamSpot, const edm::Handle<double> rho) const {
+
+
+  // Both pat and reco particles have exactly the same accessors, so we use a reco ptr
+  // throughout the code, with a single exception as of this writing, handled separately below.
+  auto superCluster = eleRecoPtr->superCluster();
+
+  // Pure ECAL -> shower shapes
+  float see            = eleRecoPtr->full5x5_sigmaIetaIeta();
+  float spp            = eleRecoPtr->full5x5_sigmaIphiIphi();
+  float OneMinusE1x5E5x5 = 1. - eleRecoPtr->full5x5_e1x5() / eleRecoPtr->full5x5_e5x5();
+  float R9             = eleRecoPtr->full5x5_r9();
+  float etawidth       = superCluster->etaWidth();
+  float phiwidth       = superCluster->phiWidth();
+  float HoE            = eleRecoPtr->full5x5_hcalOverEcal(); //hadronicOverEm();
+  // Endcap only variables
+  float PreShowerOverRaw  = superCluster->preshowerEnergy() / superCluster->rawEnergy();
+
+  // To get to CTF track information in pat::Electron, we have to have the pointer
+  // to pat::Electron, it is not accessible from the pointer to reco::GsfElectron.
+  // This behavior is reported and is expected to change in the future (post-7.4.5 some time).
+  bool validKF= false;
+  reco::TrackRef myTrackRef = eleRecoPtr->closestCtfTrackRef();
+  const pat::Electron * elePatPtr = dynamic_cast<const pat::Electron *>(eleRecoPtr);
+  // Check if this is really a pat::Electron, and if yes, get the track ref from this new
+  // pointer instead
+  if( elePatPtr != nullptr )
+    myTrackRef = elePatPtr->closestCtfTrackRef();
```

Comment:
please do not break an if statement over two lines without {}

Reply:
also wait, why do we do this differently for pat and reco electrons? If this function is properly virtual then the whole elePatPtr is not necessary (and this is good as dynamic casting can be slow).

Reply:
I lack the experience to completely understand what you mean here with the pat and reco electrons, so I didn't touch that for now. I tried however to comment out the few lines with the elePatPtr, with the result that the output for MiniAOD was not correct anymore...

Reply:
Oh sigh. The pat::Electron is improperly written. Leave it as is, I'll submit a PR to fix it but it'll touch a dataformat and it'll pull in half of CMSSW so not worth you including it for now.